taichi/util/macros.h

```diff
@@ -0,0 +1,21 @@
+#pragma once
+
+#ifndef STR
+#define _STR(...) #__VA_ARGS__
+#define STR(...) _STR(__VA_ARGS__)
+#endif
+
+#ifndef CAT
+#define _CAT(x, y) x##y
+#define CAT(x, y) _CAT(x, y)
+#endif
+
+#ifndef CAT3
+#define _CAT3(x, y, z) x##y##z
+#define CAT3(x, y, z) _CAT3(x, y, z)
+#endif
+
+#ifndef EXP
+#define _EXP(x) x
+#define EXP(x) _EXP(x)
+#endif
```

Comment:
One final thing: are these macros used anywhere?

Reply:
No, for now, but they might be used in furture, e.g.:
```
taichi/backends/opengl/shaders/atomics_data_f32.glsl.h
taichi/backends/opengl/shaders/atomics_extr_f32.glsl.h
taichi/backends/opengl/shaders/atomics_data_f64.glsl.h
taichi/backends/opengl/shaders/atomics_gtmp_f32.glsl.h
```
These shader codes are basically same, may use one `atomics_template.glsl.h` to replace them all.

Reply:
Also note that these files are only used by mesa GL, since NV has built-in atomic_float, no simulation needed.

Reply:
Sounds good. I'll leave them there then.